HopsanCore/src/ComponentUtilities/TempDirectoryHandle.cpp

```diff
@@ -0,0 +1,263 @@
+{
+    return mIsValid;
+}
+
+
+HString TempDirectoryHandle::getTempDirectory() {
+#if __cplusplus >= 201703L
+        mPath = std::filesystem::temp_directory_path().c_str();
+#else
+#ifdef _WIN32
+    char tempdirbuff[MAX_PATH+1];
+    GetTempPathA(MAX_PATH+1, tempdirbuff);
+    return HString(tempdirbuff);
+#else
+    char * val = getenv("TMPDIR");
+    if(val != nullptr) {
+        return val;
+    }
+    val = getenv("TMP");
+    if(val != nullptr) {
+        return val;
+    }
+    val = getenv("TEMP");
+    if(val != nullptr) {
+        return val;
+    }
+    val = getenv("TEMPDIR");
+    if(val != nullptr) {
+        return val;
+    }
+    return "/tmp";
+#endif
+#endif
+}
+
+HString TempDirectoryHandle::generateRandomNumericString()
+{
+#if false//__cplusplus >= 201103L
+        std::random_device rd;
+        std::default_random_engine gen(rd());
+        std::uniform_int_distribution<> distrib(0, 999999999);
+        int random = distrib(gen);
+#else
+        int random = rand() % 1000000000;
+#endif
+        std::stringstream ss;
+        ss << std::setw(6) << std::setfill('0') << random;
+        return ss.str().c_str();
+}
+
+bool TempDirectoryHandle::createDirectory(HString path)
+{
+#if __cplusplus >= 201703L
+        return std::filesystem::create_directory(mPath.c_str());
+#else
+    HString currentLevel = "";
+    std::string level;
+    if(!path.empty() && path[0] == '/') {
+        //path.erase(0,1);
+    }
+    std::stringstream ss(path.c_str());
+
+    while(std::getline(ss, level, '/')) {
+        currentLevel.append(level.c_str());
+#ifdef _WIN32
+        if (!folderExists(currentLevel) && _mkdir(currentLevel.c_str()) != 0) {
+            return false;
+        }
+#else
+        if(!currentLevel.empty()) {
+            if (!folderExists(currentLevel)) {
+                if(mkdir(currentLevel.c_str(), 0755) != 0) {
+                    return false;
+                }
+            }
+        }
+#endif
+
+        currentLevel += "/"; // don't forget to append a slash
+    }
+#endif
+    return folderExists(path);
+}
+
+
+bool TempDirectoryHandle::folderExists(HString &foldername)
+{
+    struct stat st;
+    stat(foldername.c_str(), &st);
+    return st.st_mode & S_IFDIR;
+}
+
+bool TempDirectoryHandle::removeDirectory(HString path)
+{
+    std::cout << "Trying to remove directory: " << path.c_str() << "\n";
+#if __cplusplus >= 201703L
+    std::filesystem::remove_all(mTmpPath.c_str());
+#else
+#ifdef _WIN32
+    WCHAR szDir[MAX_PATH+1];
+    SHFILEOPSTRUCTW fos = {0};
+    std::string tempStr(path.c_str());
+    std::wstring stemp = std::wstring(tempStr.begin(), tempStr.end());
+    StringCchCopyW(szDir, MAX_PATH,STRSAFE_LPCWSTR(stemp.c_str()));
+    int len = lstrlenW(szDir);
+    szDir[len+1] = 0; // Null termination
+
+    //Delete the folder including files and subfolders
+    fos.wFunc = FO_DELETE;
+    fos.pFrom = szDir;
+    fos.fFlags = FOF_NO_UI;
+    return (0 == SHFileOperationW( &fos ));
```

Comment:
Wow, wonderful Windows code :(

Reply:
We could clean up some variable names perhaps. The main problem is that the Windows API uses the WCHAR datatype. We must then convert HString -> const char* -> std::string -> std::wstring -> WCHAR, and also manually ensure the WCHAR is null terminated. The rest is basically just a function call.

Reply:
no need